tests/integration/relations/pgbouncer_provider/application-charm/src/charm.py

```diff
@@ -0,0 +1,199 @@
+    def _on_cluster1_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(f"cluster1 credentials: {event.username} {event.password}")
+        self.unit.status = ActiveStatus("received database credentials for cluster1")
+
+    def _on_cluster1_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"cluster1 endpoints have been changed to: {event.endpoints}")
+
+    def _on_cluster2_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(f"cluster2 credentials: {event.username} {event.password}")
+        self.unit.status = ActiveStatus("received database credentials for cluster2")
+
+    def _on_cluster2_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"cluster2 endpoints have been changed to: {event.endpoints}")
+
+    def _on_run_sql_action(self, event: ActionEvent):
+        logger.error(event.params)
+        relation_id = event.params["relation-id"]
+        databag = self.first_database.fetch_relation_data()[relation_id]
+        logger.error(databag)
+
+        dbname = event.params["dbname"]
+        query = event.params["query"]
+        user = databag.get("username")
+        password = databag.get("password")
+        if event.params["readonly"]:
+            host = databag.get("read-only-endpoints").split(",")[0]
+            dbname = f"{dbname}_readonly"
+        else:
+            host = databag.get("endpoints").split(",")[0]
+        endpoint = host.split(":")[0]
+        port = host.split(":")[1]
+
+        logger.error(f"running query: \n{query}")
+
+        connection = self.connect_to_database(
+            database=dbname, user=user, password=password, host=endpoint, port=port
+        )
+        cursor = connection.cursor()
+        cursor.execute(query)
+
+        # TODO consider adding a variable to expect results or not
+        try:
+            results = cursor.fetchall()
+        except psycopg2.ProgrammingError:
+            results = ["no results to fetch"]
+        logger.error(results)
+
+        event.set_results({"results": json.dumps(results)})
+
+    def connect_to_database(
+        self, database: str, user: str, host: str, password: str, port: str
+    ) -> psycopg2.extensions.connection:
+        """Creates a connection to the database.
+
+        Args:
+            database: database to connect to (defaults to the database
+                provided when the object for this class was created).
```

Comment:
It seems that there isn't a default value for this field. It requires the one that is passed to this method.